### PR TITLE
KIALI-2071 Allow the user to copy debug information

### PR DESCRIPTION
**Pull request by josejulio**
** Describe the change **

Adds a new modal screen to allow the user to copy debug information.
Currently showing redux store and the current url

** Screenshot **

peek 2019-01-22 17-20

### CODE REVIEW

**Comment by abonas** [comment]
I think that calling it "send debug information" is not the best term, because this only allows to copy it, and not actually send it.

File: src/components/Nav/HelpDropdown.tsx [hunk 1]
```diff
@@ -3,37 +3,45 @@ import { Dropdown, Icon, MenuItem } from 'patternfly-react';
...
-        />
+        <AboutUIModal ref={this.about} status={this.props.status} components={this.props.components} />
+        <DebugInformationContainer ref={this.debugInformation} />
         <Dropdown componentClass="li" id="help">
           <Dropdown.Toggle useAnchor={true} className="nav-item-iconic">
             <Icon type="pf" name="help" />
           </Dropdown.Toggle>
           <Dropdown.Menu>
-            <MenuItem onClick={() => this.about.open()}>About</MenuItem>
+            <MenuItem onClick={this.openDebugInformation}>Send debug information</MenuItem>
```
Comment (mtho11) [inline src/components/Nav/HelpDropdown.tsx]: Menu Items should start uppercase like: `Show Debug Information`

**Comment by mtho11** [comment]
Can you paste the actual output text into comment box above so we can see what it looks like?

File: package.json [hunk 2]
```diff
@@ -71,11 +71,13 @@
     "deep-freeze": "0.0.1",
     "js-yaml": "3.12.0",
     "lodash": "4.17.11",
+    "memoize-one": "5.0.0",
```
[the remark this answers is not in the document]
Comment (josejulio) [reply package.json]: IIUC `reselect` is for redux, i'm using this to  avoid computing something 3x times.

File: src/components/Nav/HelpDropdown.tsx (hunk 1 above)
Comment (josejulio) [reply src/components/Nav/HelpDropdown.tsx]: Ok

**Comment by jmazzitelli** [comment]
> Do you think `View debug information` would work better?

+1 ... but use capitalization -- "View Debug Info" (perhaps abbreviate "Information" as "Info" too?)

**Comment by josejulio** [comment]
Example output

```
Current URL: http://localhost:3000/console/overview?duration=60&pi=15000
Redux State: {"authentication":{"token":{"token":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6ImFkbWluIiwiZXhwIjoxNTQ4Mjc4NTU2fQ._KnibMH0iGFX5DTNnXvgTmG-RGBXFvmKAPvJpj_x1LI","expired_at":"2019-01-23 21:22:36.72414826 +0000 UTC m=+137692.678637638"},"username":"admin","error":false,"message":"","logged":true,"logging":false,"sessionTimeOut":1548278556766},"statusState":{"status":{"Kiali console version":"0.14.0-SNAPSHOT.3167-local-5f53075b1e087df138d44ad49f5fac874e5d57c6","Kiali core commit hash":"ed928066656f2a62eec145febfd718449fb67991","Kiali core version":"v0.14.0-SNAPSHOT","Kiali state":"running"},"components":[{"name":"Istio","version":"1.0.4"},{"name":"Prometheus","version":"2.3.1"},{"name":"Kubernetes","version":"v1.10.0+b81c8f8"},{"name":"Jaeger","url":"https://jaeger-query-istio-system.ose.os310-kprod.osoos.centralci.eng.rdu2.redhat.com"},{"name":"Grafana","url":"http://grafana-istio-system.ose.os310-kprod.osoos.centralci.eng.rdu2.redhat.com"}],"warningMessages":[]},"messageCenter":{"nextId":4,"groups":[{"id":"systemErrors","title":"Open issues","messages":[],"showActions":false,"hideIfEmpty":true},{"id":"default","title":"Notifications","messages":[{"id":0,"content":"Could not fetch namespace list, Error: [ Not Found ]","type":"error","show_notification":false,"seen":true,"created":"2019-01-23T16:18:03.598Z"},{"id":1,"content":"Could not fetch namespace list, Error: [ Not Found ]","type":"error","show_notification":false,"seen":true,"created":"2019-01-23T16:18:42.159Z"},{"id":2,"content":"Could not fetch namespace list, Error: [ Not Found ]","type":"error","show_notification":false,"seen":true,"created":"2019-01-23T16:18:54.540Z"},{"id":3,"content":"Could not fetch namespace list, Error: [ Not Found ]","type":"error","show_notification":true,"seen":false,"created":"2019-01-23T16:18:59.046Z"}],"showActions":true,"hideIfEmpty":false}],"hidden":true,"expanded":false,"expandedGroupId":"default"},"namespaces":{"activeNamespaces":[],"isFetching":false,"items":[]},"globalState":{"loadingCounter":0,"isPageVisible":true},"graph":{"cyData":null,"isLoading":false,"isError":false,"graphDataDuration":0,"graphDataTimestamp":0,"graphData":{},"layout":{"name":"dagre","fit":false,"nodeDimensionsIncludeLabels":true,"rankDir":"LR"},"summaryData":null,"filterState":{"edgeLabelMode":"hide","graphType":"versionedApp","showCircuitBreakers":true,"showFindHelp":false,"showLegend":false,"showMissingSidecars":true,"showNodeLabels":true,"showSecurity":false,"showServiceNodes":true,"showTrafficAnimation":false,"showUnusedNodes":false,"showVirtualServices":true}},"userSettings":{"interface":{"navCollapse":false},"duration":60,"refreshInterval":15000},"grafanaInfo":null,"_persist":{"version":-1,"rehydrated":true}}
```

**Comment by josejulio** [comment]
How about this:

screenshot_2019-01-23_10-21-31

**Comment by jmazzitelli** [comment]
> Example output

Is there a way we can see this output formatted/indented/prettied?

**Comment by josejulio** [comment]
> Is there a way we can see this output formatted/indented/prettied? And by that I mean the Kiali actually show the JSON formatted (you don't have to copy the formated json here in a comment :)

[yes](https://www.npmjs.com/package/json-beautify), let me work on that.

**Comment by josejulio** [comment]
WDYT?

screenshot_2019-01-23_10-39-33


```
Current URL: http://localhost:3000/console/overview?duration=60&pi=15000
Redux State: {
  "authentication": {
    "token": {
      "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6ImFkbWluIiwiZXhwIjoxNTQ4Mjc4NTU2fQ._KnibMH0iGFX5DTNnXvgTmG-RGBXFvmKAPvJpj_x1LI",
      "expired_at": "2019-01-23 21:22:36.72414826 +0000 UTC m=+137692.678637638"
    },
    "username": "admin",
    "error": false,
    "message": "",
    "logged": true,
    "logging": false,
    "sessionTimeOut": 1548278556766
  },
  "statusState": {
    "status": {
      "Kiali console version": "0.14.0-SNAPSHOT.3167-local-5f53075b1e087df138d44ad49f5fac874e5d57c6",
      "Kiali core commit hash": "ed928066656f2a62eec145febfd718449fb67991",
      "Kiali core version": "v0.14.0-SNAPSHOT",
      "Kiali state": "running"
    },
    "components": [
      {
        "name": "Istio",
        "version": "1.0.4"
      },
      {
        "name": "Prometheus",
        "version": "2.3.1"
      },
      {
        "name": "Kubernetes",
        "version": "v1.10.0+b81c8f8"
      },
      {
        "name": "Jaeger",
        "url": "https://jaeger-query-istio-system.ose.os310-kprod.osoos.centralci.eng.rdu2.redhat.com"
      },
      {
        "name": "Grafana",
        "url": "http://grafana-istio-system.ose.os310-kprod.osoos.centralci.eng.rdu2.redhat.com"
      }
    ],
    "warningMessages": []
  },
  "messageCenter": {
    "nextId": 1,
    "groups": [
      {
        "id": "systemErrors",
        "title": "Open issues",
        "messages": [],
        "showActions": false,
        "hideIfEmpty": true
      },
      {
        "id": "default",
        "title": "Notifications",
        "messages": [
          {
            "id": 0,
            "content": "Could not fetch namespace list, Error: [ Not Found ]",
            "type": "error",
            "show_notification": true,
            "seen": false,
            "created": "2019-01-23T16:39:09.534Z"
          }
        ],
        "showActions": true,
        "hideIfEmpty": false
      }
    ],
    "hidden": true,
    "expanded": false,
    "expandedGroupId": "default"
  },
  "namespaces": {
    "activeNamespaces": [],
    "isFetching": false,
    "items": []
  },
  "globalState": {
    "loadingCounter": 0,
    "isPageVisible": true
  },
  "graph": {
    "cyData": null,
    "isLoading": false,
    "isError": false,
    "graphDataDuration": 0,
    "graphDataTimestamp": 0,
    "graphData": {},
    "layout": {
      "name": "dagre",
      "fit": false,
      "nodeDimensionsIncludeLabels": true,
      "rankDir": "LR"
    },
    "summaryData": null,
    "filterState": {
      "edgeLabelMode": "hide",
      "graphType": "versionedApp",
      "showCircuitBreakers": true,
      "showFindHelp": false,
      "showLegend": false,
      "showMissingSidecars": true,
      "showNodeLabels": true,
      "showSecurity": false,
      "showServiceNodes": true,
      "showTrafficAnimation": false,
      "showUnusedNodes": false,
      "showVirtualServices": true
    }
  },
  "userSettings": {
    "interface": {
      "navCollapse": false
    },
    "duration": 60,
    "refreshInterval": 15000
  },
  "grafanaInfo": null,
  "_persist": {
    "version": -1,
    "rehydrated": true
  }
}
```

**Comment by josejulio** [comment]
I slightly modified it to have a big json that contains the currentURL and the state
screenshot_2019-01-23_10-53-27

**Comment by jmazzitelli** [comment]
This looks good. however, I don't see the graph (cy) data in the json? its null?

File: package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (mtho11) [reply package.json]: What about https://lodash.com/docs/4.17.10#memoize which we already have?

**Comment by josejulio** [comment]
> This looks good. however, I don't see the graph (cy) data in the json? its null?

I'm setting it to null in purpose. Do you think it would be useful to have it?
It has circular references and can't be directly encoded to JSON, i could use some other tool to remove the circular references if is useful.

**Comment by jmazzitelli** [comment]
> Do you think it would be useful to have it?

Initially, that was the whole point to this! :) We need a way to get the JSON that the cy graph the user is currently looking at. I wouldn't go messing around with taking out circular deps - is there no way to just put a blockquote in here or something to just put a blob of cy json in here?

**Comment by jmazzitelli** [comment]
See the title of the [JIRA 2017](https://issues.jboss.org/browse/KIALI-2071) this PR is addressing -- "Get JSON used to generate the graph"

**Comment by josejulio** [comment]
> We need a way to get the JSON that the cy graph

:) the json is what we received from the server, is in "graphData"

**Comment by jmazzitelli** [comment]
> :) the json is what we received from the server, is in "graphData"

oh, ok, so it is in there? @jshaughn this is all we need right?

**Comment by josejulio** [comment]
> oh, ok, so it is in there? @jshaughn this is all we need right?

yep, let me get you an example with actual data

**Comment by jmazzitelli** [comment]
FWIW: I also see empty graphData in the example JSON you posted:

```
"graph": {
    "cyData": null,
    "isLoading": false,
    "isError": false,
    "graphDataDuration": 0,
    "graphDataTimestamp": 0,
    "graphData": {},          <==================
...
```

**Comment by josejulio** [comment]
> FWIW: I also see empty graphData in the example JSON you posted:

Yes, I did not open the graph page in that example, I'll post one soon

**Comment by josejulio** [comment]
Full one with graph data: https://pastebin.com/DbK8FTEz

**Comment by josejulio** [comment]
Also added some new code that triggers a warning if the current output is different than the current state.

peek 2019-01-23 11-44

And an error if the auto-copy fails (for some reason like old browsers maybe?)
screenshot_2019-01-23_11-41-22

**Comment by jmazzitelli** [comment]
I haven't run this, but the screenshots all look good to me.

File: package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (josejulio) [reply package.json]: I can try that one

File: src/components/DebugInformation/DebugInformation.tsx [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+  componentDidUpdate(prevProps: DebugInformationProps, prevState: DebugInformationState) {
+    if (this.props.appState !== prevProps.appState && this.state.copyStatus === CopyStatus.COPIED) {
+      this.setState({ copyStatus: CopyStatus.OLD_COPY });
+    }
+  }
+
+  render() {
+    const renderDebugInformation = memoizeOne(() => {
+      console.log('DEBUG: I got called');
```
Comment (josejulio) [inline src/components/DebugInformation/DebugInformation.tsx]: Whoops, I 'll remove this console.log in the next commit.

File: package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (josejulio) [reply package.json]: It works, i'll remove the memoize-one

[the remark this answers is not in the document]
Comment (josejulio) [reply package.json]: done

**Pull request merged**